### PR TITLE
basecamps: Use normal recipes and fake resources

### LINKED ISSUE #30478: Base Camp Kitchen Missions do not allow Frozen ingredients to be used for some recipes
# Bug

Base Camp Kitchen Missions do not allow Frozen ingredients to be used for most recipes.

# Steps To Reproduce

1. Set frozen food (e.g. chunk of meat (frozen)) on the "Basecamp: Food" tiles so they can be accessed by NPCs.
2. Open the Base Camp Menu, Select a Kitchen Expansion, and attempt to have NPCs work with the frozen ingredients.
3. Most recipes will be grey/invalid, as they seem unable to use the frozen versions of the ingredient.

# Expected behavior

NPCs should be allowed to use frozen ingredients in the Base Camp's Kitchen Expansion via cooking missions.

My testing has shown the following recipes are impacted:

Cook: Cornmeal
Cook: Egg, Boiled
Cook: Fish, Cooked
Cook: Fish, Smoked
Cook: Lard
Cook: Meat, Aspic
Cook: Meat Jerky
Cook: Meat Pie
Cook: Meat, Cooked
Cook: Meat, Smoked
Cook: Mushroom, Cooked
Cook: Mushroom, Dried
Cook: Sausage
Cook: Sausage, Wasteland
Cook: Veggy Aspic
Cook: Veggy Pie
Cook: Veggy, Pickled

The following recipes are not impacted, or seem to work with (some) frozen ingredients:

Cook: Hardtack (frozen flour works OK, but frozen salt does NOT, so it technically cant be made)
Cook: Bread
Cook: Flatbread
Cook: Meat, Canned

(Unlisted/missing recipes were ones I was unable to test at the time.)

# Screenshots

Here you can see a stack of frozen chunks of meat placed on the Basecamp: Food tile for usage

Here you can see the Cook: Meat, Canned is able to use the ~1600 peices of meat as expected 

Here you can see both the Cook: Meat, Cooked + Cook: Meat, Smoked are unable to pull from this same pile of (frozen) chunks of meat

# Versions and configuration
- OS: Windows [64-bit]
- Game Version: 0.D-2863-gca633a0
- Graphics Version: Tiles
- Mods loaded: 
    Dark Days Ahead [dda],
    Urban Development [Urban_Development],
    More Locations [more_locations],
    Vehicle Additions Pack [blazemod],
    Tanks and Other Vehicles [Tanks],
    Boats [boats],
    Folding Parts pack [deoxymod],
    Filthy Clothing [filthy_morale],
    StatsThroughSkills [StatsThroughSkills],
    Classes and Scenarios Mod [more_classes_scenarios],
    More Survival Tools [More_Survival_Tools],
    Makeshift Items Mod [makeshift],
    Extended Realistic Guns [realguns],
    No Fictional Guns [no_rivtech_guns],
    Mutant NPCs [mutant_npcs],
    Mining Mod [Mining_Mod],
    NPC traits [npc_traits],
    Modular Turrets [modular_turrets],
    Salvaged Robots [Salvaged_Robots],
    [DP_MOD] Indicators: Chesthole [DP_AMAZING_CHEST_INDICATORS],
    [DP_MOD] No Highlight [DP_No_Highlight]

(Am new to GitHub, so constructive criticism welcome for next time)

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "basecamps: Use normal recipes and fake resources"```

#### Purpose of change
Fixes #30478
Basecamps currently have their own set of recipes that don't batch correctly and are guaranteed to drift out of sync with normal recipes over time.  Also, upgrading a basecamp creates all kinds of tools and furniture that cannot be easily used.

Address both issues by revising basecamp upgrade recipes to provide fake items to the camp inventory which can then  be used for basecamp crafting, and then revising camp recipes to just be normal recipes that can take advantage of fake camp items.

#### Describe the solution
Includes #30469
Add an optional `basecamp *` argument to use_charges(), and when then option is present, call into a basecamp::use_charges() function.  This is similar to the way that vehicle part pseudo items consume charges in the current map::use_charges, but with slightly better encapsulation.

Add a new blueprint_resources field to recipes, which is a list of fake items that the recipe provides if it is a blueprint.

Add an _inv inventory field to basecamp::, and populate it *once* at the start of generating the mission_key.  Parse the list of resources provided by the basecamp and add the fake items with appropriate charges to the _inv.  Crafting requirements can be resolved by the fake items, and charges consumed, appropriately.

Add the fake items and their blueprint resources to data/json.

Finally, revise camp recipes to almost always use standard recipes.

#### Describe alternatives you've considered
I'm doing lazy evaluation of camp resources based on the camp's completed upgrade recipes.  The really correct solution would be add resources based on the camp's map, and have those get automatically updated on ter_set()/furn_set(), which is the long term goal.  This is a good intermediate step.

Camp recipe groups are currently kind of backwards: recipes are placed in groups, instead of groups being created on the fly from tagged recipes.  That's a relatively straightforward change that I'll make later as this PR is large enough as it is.

### CODE REVIEW

File: src/basecamp.cpp [hunk 1]
```diff
@@ -328,51 +360,126 @@ void basecamp::set_name( const std::string &new_name )
     name = new_name;
 }
 
-void basecamp::consume_components( inventory &camp_inv, const recipe &making, int batch_size,
+/*
+ * we could put this logic in map::use_charges() the way the vehicle code does, but I think
+ * that's sloppy
+ */
+long basecamp::use_charges( const itype_id fake_id, long quantity, std::list<item> &ret )
```
Comment (ifreund) [inline src/basecamp.cpp]: `fake_id` could be passed by reference.

Also, is there a good reason not to follow the convention set by `map::use_charges()` of passing `quantity` by reference and returning an `std::list<item>`?

**Comment by davidpwbrown** [comment]
With this, I might try my hand at making a new mission to transport resources from one camp to another. That may become more useful as they continue to expand.
This is very useful, thanks!

File: src/map.cpp [hunk 2]
```diff
@@ -4949,6 +4950,13 @@ std::list<item> map::use_charges( const tripoint &origin, const int range,
         }
     }
 
+    if( bcp ) {
+        quantity = bcp->use_charges( type, quantity, ret );
+        if( quantity == 0 ) {
```
Comment (ifreund) [inline src/map.cpp]: Probably better to make this `<=` like other places in this function for safety if something goes wrong.
```suggestion
        if( quantity <= 0 ) {
```

File: src/item_factory.cpp [hunk 3]
```diff
@@ -1492,6 +1492,7 @@ void Item_factory::load( islot_tool &slot, JsonObject &jo, const std::string &sr
     assign( jo, "initial_charges", slot.def_charges, strict, 0L );
     assign( jo, "charges_per_use", slot.charges_per_use, strict,
             static_cast<decltype( slot.charges_per_use )>( 0 ) );
+    assign( jo, "charge_factor", slot.charge_factor, strict, 1 );
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply src/item_factory.cpp]: https://github.com/CleverRaven/Cataclysm-DDA/blob/4da91581d1e76ad21b90d94d136e8876f5a0d688/src/item_factory.cpp#L1499

```
  {
    "id": "fake_fireplace",
    "type": "TOOL",
    "copy-from": "fake_item",
    "name": "basecamp fireplace",
    "description": "A fake fireplace used for basecamps.",
    "sub": "hotplate",
    "ammo": "tinder",
    "max_charges": 50000,
    "charge_factor": 25
  },
```

I'm not sure you mean about not loading sub from the JSON.  Tools load "sub" already, the fake_fireplace is a TOOL, it has `"sub": "hotplate"`, and then I'm using "charge_factor" to convert 1 charge of hotplate batteries to 25 charges of tinder.

[the remark this answers is not in the document]
Comment (ifreund) [reply src/item_factory.cpp]: Sorry, I got it into my head that sub was a new thing you just added, which of course it isn't.  There's no problem here

File: src/basecamp.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/basecamp.cpp]: Yes, the `use_charges_from_furn()` syntax that doesn't require a splice() is better than the `use_charges_from_stack()` syntax that does.

But it doesn't matter in this instance so I'll return the list.

**Comment by kevingranade** [comment]
Conflicted and IDK how much work it'll be to sort out.

**Pull request merged**